Update install script, add missing dependencies, and various platform-specific tweaks

dfilaretti opened the pull request:
Working:

* Ubuntu 
* Debian 
* Alpine Linux
* Arch Linux
* Fedora
* Darwin

Not working:

* CentOS

dfilaretti commented on PLATFORMS_INSTALL.md:
```diff
@@ -0,0 +1,56 @@
+
+The standard installation steps apply, with this single exception:
+
+* Use `cmake -DCMAKE_PREFIX_PATH=/usr/local/opt/llvm ..`
```
Hmm no, I think `cmake` gave the usual error it gives when no `llvm` is found. I installed it via `brew`: `brew install llvm`

dfilaretti commented on scripts/install_deps.sh:
```diff
@@ -186,36 +190,42 @@ case $(uname -s) in
                         #buster
                         echo "Installing solidity dependencies on Debian Buster (10.x)."
                         install_z3="libz3-dev"
+                        # llvm5 not available here??
```
Even if it's still the "testing" version? (https://wiki.debian.org/DebianBuster)

This is where I'm getting the Debian packages from: https://apt.llvm.org/ - it looks like there's an "unstable" section (even if it's not called "Buster"). Maybe that will work. I'll try.

dwightguth replied:
looks like stretch and buster come with llvm-5.0 in their regular package repo, so you don't need to download from apt.llvm.org.

dfilaretti replied:
How/where did you find it? When I tried installing it on stretch I thought it was using version 4 or something like that. And when I did search online, it appeared to me that llvm5 was coming from some "unstable" package repo (e.g. https://packages.debian.org/search?keywords=llvm-5.0)

dwightguth replied:
https://packages.debian.org/stretch-backports/llvm-5.0

It's in backports.

dfilaretti replied:
OK, but they also need to be added separately to `sources.list` (https://backports.debian.org/Instructions/). Will do that then.

dfilaretti commented on PLATFORMS_INSTALL.md:
```diff
@@ -0,0 +1,56 @@
+* `Cmake` error
+   * Requires manually changing `llvm` to `llvm4` in `/usr/lib/cmake/llvm5/LLVMConfig.cmake`
+* After that, still have several compilation issues 
+   * `/home/vagrant/solidity/libiele/IeleValue.h:3:34: fatal error: llvm/Support/Casting.h: No such file or directory
```
Where should this option go? I tried make -I /usr/include/llvm4 but nothing changes apparently.

dwightguth replied:
it needs to be passed to the C compiler.

dfilaretti replied:
Thank you, this seems to work.

dfilaretti commented:
Seems those 3 failures are in `sol2iele` too...

dfilaretti commented:
@dwightguth I addressed your comments/suggestions and was able to remove and simplify a lot of the code. Alpine 3.6 now works. 3.7 is still broken (see `PLATFORM_INSTALL.md`) as it seems the `llvm` package is not installed correctly. CentOS is not working too. 

There's one one more thing though, which I asked yesterday on Slack but received no reply: in `EthDependencies.cmake` we currently have `find_package(LLVM 5.0.0 QUIET REQUIRED CONFIG)`; this requires version `5` only, so we want to relax it so that it can be version `4` or greater. I am not sure how to do that - looked online but couldn't find what I needed. Notice that we could remove the `5.0.0` from the line above and it will work on a fresh install; howevr, it may fail in case there are multiple versions of `llvm` installed - e.g. on Jenkins, it ends up using version `3.6`.

dfilaretti commented:
Hmm so there's a way of telling cmake to pick the greatest version (of llvm),  in case it can find more than one; this would fix Jenkins failure (since it would not pick `3.9` by default anymore). 
* https://stackoverflow.com/a/44970744
* https://cmake.org/cmake/help/v3.9/variable/CMAKE_FIND_PACKAGE_SORT_ORDER.html#variable:CMAKE_FIND_PACKAGE_SORT_ORDER

However this seems to require cmake version >=3.7, while `3.5` is installed.

dfilaretti commented:
Problem above is solved.

dwightguth commented on cmake/EthCompilerSettings.cmake:
```diff
@@ -27,7 +27,11 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
+	
+	# NOTE: this turned out to be necessary in only one case, Alpine Linux 3.6. 
+	#       Not sure whether this is the right place for this?
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I /usr/include/llvm4")
```
you should put this in EthDependencies.cmake inside the block for llvm 4 I think...

dfilaretti replied:
Yeah that would make sense, but unfortunately it seems it's not going to work for... I am leaving it as it is now, perhaps should investigate later on...

dwightguth commented on scripts/install_deps.sh:
```diff
@@ -186,11 +189,13 @@ case $(uname -s) in
                     8)
                         #jessie
                         echo "Installing solidity dependencies on Debian Jesse (8.x)."
+                        echo "deb http://ftp.debian.org/debian jessie-backports main" | sudo tee /etc/apt/sources.list.d/jessie-backports.list
```
I'm not sure jessie-backports has llvm 5, did you test it?

dwightguth commented:
@theo25 this is ready for review. CentOS isn't working yet, but as far as I can tell the code that exists on mainline solidity for CentOS doesn't really work very well either, and I'm hesitant to try to modify something so obviously a ball of kludge. I think we can address this at a later date if users request it.

theo25 commented on libsolidity/inlineasm/AsmParser.cpp:
```diff
@@ -606,7 +606,7 @@ bool Parser::isValidNumberLiteral(string const& _literal)
 {
 	try
 	{
-		u256(_literal);
```
why can't we just remove the whole try/catch block?

dwightguth replied:
well we could I suppose but this is code we don't actually need to support since it's specific to inline assembly, and this was the only line that generated a compiler warning, so I just commented out that one line...

theo25 replied:
OK that's fine then
